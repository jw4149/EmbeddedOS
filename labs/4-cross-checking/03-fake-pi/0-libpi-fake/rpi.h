#ifndef __RPI_H__
#define __RPI_H__
// simplest way to rewrap --- just define replacements in header --- don't
// even need .c
//
// easy, but two downsides: 
//   1. we are polluting the pi program namespace with system includes.
//   2. we aren't using the raw rpi.h so can get different behavior.
#include <stdio.h>
#include <stdlib.h>

// redefine printk as fprintf using variadic macros.
#define printk(args...) fprintf(stderr, args)

static void inline clean_reboot(void) {
    printk("DONE!!!\n");
    exit(0);
}

void notmain(void);
#endif
