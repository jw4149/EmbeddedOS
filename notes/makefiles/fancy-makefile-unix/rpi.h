#include <stdio.h>

#define printk printf
#define debug printf
