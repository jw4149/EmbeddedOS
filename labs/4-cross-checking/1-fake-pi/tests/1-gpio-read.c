#include "rpi.h"
void notmain(void) {
    output("about to do a single read from pins [0..32)\n");
    for(unsigned pin = 0; pin < 32; pin++) {
        output("reading from pin %d\n", pin);
        gpio_read(pin);
    }
}
