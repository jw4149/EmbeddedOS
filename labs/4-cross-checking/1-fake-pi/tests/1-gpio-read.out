TRACE:0: calling pi code
TRACE:1: GET32(0x20200034) = 0x625558ec
TRACE:2: DEV_VAL32=0x0
TRACE:3: GET32(0x20200034) = 0x238e1f29
TRACE:4: DEV_VAL32=0x0
TRACE:5: GET32(0x20200034) = 0x46e87ccd
TRACE:6: DEV_VAL32=0x1
TRACE:7: GET32(0x20200034) = 0x3d1b58ba
TRACE:8: DEV_VAL32=0x1
TRACE:9: GET32(0x20200034) = 0x507ed7ab
TRACE:10: DEV_VAL32=0x0
TRACE:11: GET32(0x20200034) = 0x2eb141f2
TRACE:12: DEV_VAL32=0x1
TRACE:13: GET32(0x20200034) = 0x41b71efb
TRACE:14: DEV_VAL32=0x1
TRACE:15: GET32(0x20200034) = 0x79e2a9e3
TRACE:16: DEV_VAL32=0x1
TRACE:17: GET32(0x20200034) = 0x7545e146
TRACE:18: DEV_VAL32=0x1
TRACE:19: GET32(0x20200034) = 0x515f007c
TRACE:20: DEV_VAL32=0x0
TRACE:21: GET32(0x20200034) = 0x5bd062c2
TRACE:22: DEV_VAL32=0x0
TRACE:23: GET32(0x20200034) = 0x12200854
TRACE:24: DEV_VAL32=0x1
TRACE:25: GET32(0x20200034) = 0x4db127f8
TRACE:26: DEV_VAL32=0x0
TRACE:27: GET32(0x20200034) = 0x216231b
TRACE:28: DEV_VAL32=0x1
TRACE:29: GET32(0x20200034) = 0x1f16e9e8
TRACE:30: DEV_VAL32=0x1
TRACE:31: GET32(0x20200034) = 0x1190cde7
TRACE:32: DEV_VAL32=0x1
TRACE:33: GET32(0x20200034) = 0x66ef438d
TRACE:34: DEV_VAL32=0x1
TRACE:35: GET32(0x20200034) = 0x140e0f76
TRACE:36: DEV_VAL32=0x1
TRACE:37: GET32(0x20200034) = 0x3352255a
TRACE:38: DEV_VAL32=0x0
TRACE:39: GET32(0x20200034) = 0x109cf92e
TRACE:40: DEV_VAL32=0x1
TRACE:41: GET32(0x20200034) = 0xded7263
TRACE:42: DEV_VAL32=0x0
TRACE:43: GET32(0x20200034) = 0x7fdcc233
TRACE:44: DEV_VAL32=0x0
TRACE:45: GET32(0x20200034) = 0x1befd79f
TRACE:46: DEV_VAL32=0x1
TRACE:47: GET32(0x20200034) = 0x41a7c4c9
TRACE:48: DEV_VAL32=0x1
TRACE:49: GET32(0x20200034) = 0x6b68079a
TRACE:50: DEV_VAL32=0x1
TRACE:51: GET32(0x20200034) = 0x4e6afb66
TRACE:52: DEV_VAL32=0x1
TRACE:53: GET32(0x20200034) = 0x25e45d32
TRACE:54: DEV_VAL32=0x1
TRACE:55: GET32(0x20200034) = 0x519b500d
TRACE:56: DEV_VAL32=0x0
TRACE:57: GET32(0x20200034) = 0x431bd7b7
TRACE:58: DEV_VAL32=0x0
TRACE:59: GET32(0x20200034) = 0x3f2dba31
TRACE:60: DEV_VAL32=0x1
TRACE:61: GET32(0x20200034) = 0x7c83e458
TRACE:62: DEV_VAL32=0x1
TRACE:63: GET32(0x20200034) = 0x257130a3
TRACE:64: DEV_VAL32=0x0
TRACE: pi exited cleanly: 39 calls to random
